### PR TITLE
WebRTC data channels backend

### PR DESCRIPTION
**Pull request by ferjm**
This is not entirely spec compliant, but it contains the basic stuff to make data channels work. It allows sending only strings for now.

### CODE REVIEW

File: backends/gstreamer/datachannel.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                    .expect("Invalid data channel message")
+                {
+                    callbacks_.lock().unwrap().message(message);
+                }
+                None
+            })
+            .map_err(|e| e.to_string())?;
+
+        let channel = InnerDataChannel {
+            channel: glib::SendWeakRef::from(channel.downgrade()),
```
Comment (sdroege) [inline backends/gstreamer/datachannel.rs:105]: This is going to explode sooner or later. The datachannel object is thread-safe, but until https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/issues/1168 is solved you'll have to put your own wrapper struct around it on which `Send`/`Sync` are implemented.

File: backends/gstreamer/datachannel.rs [hunk 2]
```diff
@@ -129,15 +111,11 @@ impl WebRtcDataChannelBackend for GStreamerWebRtcDataChannel {
...
-        // was created on the webrtc controller's thread.
-        self.webrtc_thread
-            .internal_event(InternalEvent::SendDataChannelMessage(
-                Box::new(self.channel.clone()),
-                message.to_owned(),
-            ));
-        Ok(())
+        self.channel
+            .0
+            .emit("send-string", &[&Value::from(message)])
```
Comment (sdroege) [inline backends/gstreamer/datachannel.rs:116]: While at it, you could also hide the signal emission things in functions inside `impl DataChannel` :)

File: backends/gstreamer/datachannel.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            })
+            .map_err(|e| e.to_string())?;
+
+        let channel = glib::SendWeakRef::from(channel.downgrade());
+        Ok(Self { channel, callbacks })
+    }
+}
+
+impl WebRtcDataChannel for GStreamerWebRtcDataChannel {
+    fn set_on_open(&self, cb: SendBoxFnOnce<'static, ()>) {
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply backends/gstreamer/datachannel.rs]: `send()` can be called by the controller instead

File: webrtc/lib.rs [hunk 4]
```diff
@@ -50,11 +52,10 @@ pub trait WebRtcControllerBackend: Send {
     fn create_offer(&mut self, cb: SendBoxFnOnce<'static, (SessionDescription,)>) -> WebRtcResult;
     fn create_answer(&mut self, cb: SendBoxFnOnce<'static, (SessionDescription,)>) -> WebRtcResult;
     fn add_stream(&mut self, stream: &MediaStreamId) -> WebRtcResult;
-    fn create_data_channel(
-        &mut self,
-        init: &WebRtcDataChannelInit,
-        channel: Sender<Box<dyn WebRtcDataChannelBackend>>,
-    ) -> WebRtcResult;
+
+    fn create_data_channel(&mut self, id: &DataChannelId, init: &DataChannelInit) -> WebRtcResult;
```
Comment (Manishearth) [inline webrtc/lib.rs:56]: the controller probably can manage ids itself and return an id here.

File: backends/gstreamer/datachannel.rs [hunk 5]
```diff
@@ -45,83 +57,96 @@ impl GStreamerWebRtcDataChannel {
...
+                            DataChannelEvent::Error(WebRtcError::Backend(error.to_string())),
+                        ));
                 }
                 None
             })
             .map_err(|e| e.to_string())?;
 
-        let callbacks_ = callbacks.clone();
+        let id_ = id.clone();
+        let thread_ = Mutex::new(thread.clone());
```
Comment (Manishearth) [inline backends/gstreamer/datachannel.rs:119]: the outer mutex doesn't seem necessary?

Comment (Manishearth) [reply backends/gstreamer/datachannel.rs:119]: @sdroege should the `on-message-string` and `on-error` callbacks require Sync here? I recall we fixed some of these cases earlier: https://github.com/sdroege/gstreamer-rs/issues/154

File: webrtc/datachannel.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
+use crate::{WebRtcError, WebRtcResult};
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use uuid::Uuid;
+
+lazy_static! {
```
Comment (Manishearth) [inline webrtc/datachannel.rs:7]: I think these ids should just be numbers and individual webrtc implementations should produce and consume them

We shouldn't have a `DataChannelBackend` trait at all, all of this should be internal to the webrtc impl

File: examples/simple_webrtc.rs [hunk 7]
```diff
@@ -214,10 +198,30 @@ impl WebRtcSignaller for Signaller {
...
-        }));
+    fn on_data_channel_event(
+        &self,
+        id: DataChannelId,
+        event: DataChannelEvent,
+        controller: &WebRtcController,
+    ) {
+        match event {
+            DataChannelEvent::NewChannel => {
+                println!("New channel {:?}", id);
```
Comment (Manishearth) [inline examples/simple_webrtc.rs:209]: probably should be removed or turned into `debug!()`

File: backends/gstreamer/datachannel.rs (hunk 5 above)
Comment (ferjm) [reply backends/gstreamer/datachannel.rs:119]: `rustc` complains if I don't add the outer mutex

Comment (Manishearth) [reply backends/gstreamer/datachannel.rs:119]: Yeah sorry this is the Sync requirement that I think needs to be removed upstream. No worries for now

**Pull request merged**